src/library/scala/math/BigInt.scala

```diff
@@ -108,14 +109,56 @@ object BigInt {
   /** Implicit conversion from `java.math.BigInteger` to `scala.BigInt`.
    */
   implicit def javaBigInteger2bigInt(x: BigInteger): BigInt = apply(x)
+
+  /** Computes the GCD of the two positive Long arguments, according to the binary GCD algorithm.
+   * Taken from the Spire library. */
+  private def longGcd(_x: Long, _y: Long): Long = {
+    if (_x == 1L || _y == 1L) return 1L
+
+    var x = _x
+    val xz = java.lang.Long.numberOfTrailingZeros(x)
+    x >>= xz
+
+    var y = _y
+    val yz = java.lang.Long.numberOfTrailingZeros(y)
+    y >>= yz
+
+    while (x != y) {
+      if (x > y) {
+        x -= y
+        x >>= java.lang.Long.numberOfTrailingZeros(x)
+      } else {
+        y -= x
+        y >>= java.lang.Long.numberOfTrailingZeros(y)
+      }
+    }
+
+    if (xz < yz) x << xz else x << yz
+  }
+
 }
 
-final class BigInt(val bigInteger: BigInteger)
+/** A type with efficient encoding of arbitrary integers.
+ *
+ * It wraps `java.math.BigInteger`, with optimization for small values that can be encoded in a `Long`.
+ */
+final class BigInt private (private val _bigInteger: BigInteger, private val _long: Long)
   extends ScalaNumber
     with ScalaNumericConversions
     with Serializable
     with Ordered[BigInt]
 {
+  // Class invariant: if the number fits in a Long, then _bigInteger is null and the number is stored in _long
+  //                  otherwise, _bigInteger stores the number and _long = 0L
+  def this(_bigInteger: BigInteger) = this(
+    if (_bigInteger.bitLength <= 63) null else _bigInteger,
+    if (_bigInteger.bitLength <= 63) _bigInteger.longValue else 0L
```

Comment:
I'd love to! What can I do in a secondary constructor to achieve that?

Reply:
Why even bother storing the long version if we explicitly put in a `BigInteger`?

Reply:
We don't store a `BigInteger` in case the value fits in a `Long`: hence the test of the argument bit length.

This constructor has two purposes:

1. It preserves binary compatibility: `BigInt` had a primary constructor taking a `BigInteger` value. Here we store that value efficiently. Of course, a call to the `bigInteger` member is wasteful, because it reallocates a `BigInteger`. However, the purpose of the redesign is to avoid operating on `BigInteger`s whenever possible. If that constructor was not public, this logic would have gone in the companion object method `apply`, but this is not possible.

2. Sometimes, the result of an operation involving two `BigInteger`s will result in an integer that fits in a long. This method will make sure that the class invariant is preserved in that case.

Reply:
@Ichoran The different types of encoding have been clarified in the latest commit.

Reply:
@johnynek The new encoding takes care of reading `bitLength` once.

@Ichoran Happy with the clarifications?

Reply:
ping @Ichoran